patmat_wrappers/result.py

```diff
@@ -0,0 +1,255 @@
+
+    def ok(self) -> Option:
+        return Some(self.Value)
+
+    def err(self) -> Option:
+        return Empty()
+
+    def map(self, f: Callable[[T], U]) -> Result:
+        return Ok(f(self.Value))
+
+    def map_or(self, default: U, f: Callable[[T], U]) -> U:
+        return self.map(f)
+
+    def map_or_else(self, default: Callable[[E], U], f: Callable[[T], U]) -> U:
+        return self.map(f)
+
+    def map_err(self, f: Callable[[E], U]) -> Result:
+        return self
+
+    def iter(self) -> Iterator[T]:
+        def _iter():
+            yield self.Value
+        return iter(_iter())
+
+    def _and(self, res: Result) -> Result:
+        return res
+
+    def and_then(self, op: Callable[[T], Result]) -> Result:
+        return op(self.Value)
+
+    def _or(self, res: Result) -> Result:
+        return self
+
+    def or_else(self, op: Callable[[E], U]) -> Result:
+        return self
+
+    def unwrap(self) -> T:
+        return self.Value
+
+    def unwrap_or(self, default: T) -> T:
+        return self.Value
+
+    def unwrap_or_else(self, default: Callable[[], T]) -> T:
+        return self.Value
+
+    def expect(self, msg: str) -> T:
+        return self.Value
+
+    def unwrap_err(self) -> E:
+        raise UnwrapException(f"{self.Value}")
+
+    def expect_err(self, msg: str) -> E:
+        raise UnwrapException(msg)
+
+
+@dataclass
+class Err(ResultProtocol):
+    Error: E
+
+    @property
+    def is_ok(self) -> bool:
+        return False
+
+    @property
+    def is_err(self) -> bool:
+        return True
+
+    def contains(self, value: T) -> bool:
+        return False
+
+    def contains_err(self, err: E) -> bool:
+        return self.Error == err
+
+    def ok(self) -> Option:
+        return Empty()
+
+    def err(self) -> Option:
+        return Some(self.Error)
+
+    def map(self, f: Callable[[T], U]) -> Result:
+        return self
+
+    def map_or(self, default: U, f: Callable[[T], U]) -> U:
+        return default
+
+    def map_or_else(self, default: Callable[[E], U], f: Callable[[T], U]) -> U:
+        return default(self.Error)
+
+    def map_err(self, f: Callable[[E], U]) -> Result:
+        return Err(f(self.Error))
+
+    def iter(self) -> Iterator[T]:
+        return iter(e for e in tuple())
+
+    def _and(self, res: Result) -> Result:
+        return self
+
+    def and_then(self, op: Callable[[T], Result]) -> Result:
+        return self
+
+    def _or(self, res: Result) -> Result:
+        return res
+
+    def or_else(self, op: Callable[[E], U]) -> Result:
+        return Err(op(self.Error))
+
+    def unwrap(self) -> T:
+        raise UnwrapException(self.Error)
+
+    def unwrap_or(self, default: T) -> T:
+        return default
+
+    def unwrap_or_else(self, default: Callable[[], T]) -> T:
+        return default()
+
+    def expect(self, msg: str) -> T:
+        raise UnwrapException(msg)
+
+    def expect_err(self, msg: str) -> E:
+        return self.Error
+
+    def unwrap_err(self) -> E:
```

Comment:
Keep the order, it's not necessary if you don't want to, but it's nice to have it.
Ok class has unwrap_err and expect_err and this one has the other way around

Reply:
Yep, auto fill didn't work as expected.